### ISSUE TITLE
NFSP multi-discrete reservoir support (blocking #126 follow-up — bucket-brigade NFSP convergence test)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#127]
## Background

Surfaced during PR #126 (closes #120). \`NfspTrainer::train_average_policies\` stores only \`row[0]\` from the BR action tensor and reshapes as \`[mb, 1]\`, which works for \`MlpBurnPolicy\` (single discrete action) but causes a Burn \`Squeeze\` panic when paired with \`MultiDiscreteMlpBurnPolicy\` (action shape \`Vec<usize>\`, e.g. bucket-brigade's \`[10, 2, 2]\`).

PR #126 sidesteps the limitation by wrapping bucket-brigade's factored action shape as a Cartesian-product \`Discrete(40)\` (10×2×2) with \`MlpBurnPolicy\`. That works structurally but expands the action space combinatorially — for envs with larger factored action shapes (think \`[10, 10, 10]\` = 1000 actions) the Cartesian product becomes intractable.

## Scope

1. Audit \`src/multi_agent/nfsp.rs\` — specifically \`train_average_policies\` and any other site that assumes scalar-action reservoirs.
2. Generalize the reservoir storage and average-policy supervised-training loop to handle \`Vec<usize>\` actions. The reservoir buffer (Vitter Algorithm R) stays — only the action storage shape changes.
3. \`MultiDiscreteMlpBurnPolicy::evaluate_actions_joint\` (verify signature) and \`MultiDiscreteMlpBurnPolicy::supervised_loss\` (or equivalent) consume \`Tensor<B, 2, Int>\` of shape \`[batch_size, num_action_dims]\`.
4. Unit tests parametrized over both single-discrete (\`MlpBurnPolicy\`) and multi-discrete (\`MultiDiscreteMlpBurnPolicy\`) action shapes.
5. Replace PR #126's \`SingleDiscreteBucketBrigade\` Cartesian-product wrapper with the native multi-discrete code path; verify \`tests/test_nfsp_bucket_brigade.rs\` still passes after the swap.

## Out of scope

- Cell-specific gap_closed baselines (separate follow-up — see companion issue).
- PSRO multi-discrete support (PSRO does not consume an average policy; only the per-agent best-response training does, and that already routes through \`JointMultiAgentTrainer\` which supports multi-discrete via PR #103).

## Acceptance criteria sketch (Curator should refine)

- [ ] NFSP reservoir storage generalized to \`Vec<usize>\` actions.
- [ ] Supervised average-policy training loop accepts multi-discrete actions.
- [ ] \`tests/test_nfsp_n_player_matching_pennies.rs\` (single-discrete, regression bar) passes unchanged.
- [ ] New \`tests/test_nfsp_multi_discrete_smoke.rs\` exercising a small multi-discrete env (or a parametrized test variant).
- [ ] PR #126's \`SingleDiscreteBucketBrigade\` wrapper removed; \`tests/test_nfsp_bucket_brigade.rs\` uses the factored action shape directly.
- [ ] \`cargo clippy --all-targets --features training -- -D warnings\` clean.

## Source

Surfaced during PR #126 implementation. Builder report: "NFSP multi-discrete reservoirs are not implemented. \`NfspTrainer::train_average_policies\` stores only \`row[0]\` and reshapes as \`[mb, 1]\`, which causes a Burn \`Squeeze\` panic for \`MultiDiscreteMlpBurnPolicy\`."

\`tier:goal-supporting\` — completes NFSP's multi-action-shape support for any env beyond single-discrete.

**Comment by rjwalters** [comment]
## Curator notes (2026-06-16)

Light-to-medium curation against `origin/main` HEAD `83d40db` (post #131). Verified the failure mode is exactly as the issue body describes and surfaced a few concrete cleavages for the Builder.

### 1. Verified failure mode at the exact line

`NfspTrainer::train_average_policies` lives at `src/multi_agent/nfsp.rs:774` (line drift since the issue body was written). The breaking sequence inside the per-agent supervised loop:

- `src/multi_agent/nfsp.rs:799-807` — flatten the reservoir batch into `obs_flat: Vec<f32>` plus `acts: Vec<i64>` (one scalar per row).
- `src/multi_agent/nfsp.rs:812-813` — build `acts_t: Tensor<B, 1, Int>` of shape `[mb]`.
- `src/multi_agent/nfsp.rs:818` — `let acts_2d = acts_t.unsqueeze_dim::<2>(1);` produces `[mb, 1]`.
- `src/multi_agent/nfsp.rs:825` — `policy.evaluate_actions_joint(obs_t, acts_2d)` dispatches through the `JointPolicy` impl.

Panic site for `MultiDiscreteMlpBurnPolicy` (action_dims `[10, 2, 2]`):

- `src/multi_agent/joint.rs:203-209` — multi-discrete impl forwards `actions` (shape `[mb, 1]`) unchanged to `evaluate_actions`.
- `src/policy/multi_discrete_mlp.rs:266-286` — loop iterates `i` from 0 to `num_dims-1 = 2`. At `i=1`, `actions.clone().slice([0..mb, 1..2])` reads column index 1 of a tensor whose column dim is 1 — out of bounds. Burn surfaces this via the subsequent `.squeeze_dim::<1>(1)` on a malformed slice. Effective panic site: line 274.

(For `MlpBurnPolicy`, the `JointPolicy` impl at `src/multi_agent/joint.rs:167-176` defensively squeezes `[mb, 1]` to `[mb]`, so the single-discrete path is unaffected.)

### 2. Other NFSP sites with single-discrete assumptions

Audited the full `nfsp.rs` for `i64`-as-scalar assumptions:

- `src/multi_agent/nfsp.rs:381` — `reservoirs: Vec<ReservoirBuffer<(Vec<f32>, i64)>>` (storage type — needs generalization).
- `src/multi_agent/nfsp.rs:489` — public accessor `pub fn reservoir(&self, i: usize) -> &ReservoirBuffer<(Vec<f32>, i64)>` (API surface — needs the same type swap).
- `src/multi_agent/nfsp.rs:709` — `self.reservoirs[i].push((agent_obs.clone(), row[0]))` — single-action push.
- `src/multi_agent/nfsp.rs:799-807` — `train_average_policies` flatten loop assumes a scalar `a`.
- `src/multi_agent/nfsp.rs:812-818` — `Tensor<B, 1, Int>` of shape `[mb]` plus `unsqueeze_dim::<2>(1)` — both shapes need to change to `[mb, num_action_dims]`.

The rollout collector at `src/multi_agent/nfsp.rs:646-768` already handles multi-dim actions correctly: it collects per-dim slices into `row: Vec<i64>` of length `num_action_dims` (line 702-720), strides into `act_buf` correctly (line 722-723), and pushes the full vector into `joint_action` (line 724). Only the reservoir push at 709 truncates to `row[0]`. So the rollout side is fine — the work is entirely in the reservoir/AP-supervised path.

`ReservoirBuffer` itself (lines 113-232) is generic over `T: Clone` already — it has no scalar-action assumption. The change is purely in how `NfspTrainer` parameterizes that `T`.

### 3. `MultiDiscreteMlpBurnPolicy` API surface

- `evaluate_actions_joint`: `src/multi_agent/joint.rs:203-209` forwards to `MultiDiscreteMlpBurnPolicy::evaluate_actions` at `src/policy/multi_discrete_mlp.rs:253-293`.
- Expected actions shape: `[batch, num_action_dims]` (one int column per action dim).
- Returns: `(log_probs: Tensor<B, 1>, entropy: Tensor<B, 1>, values: Tensor<B, 1>)` where `log_probs` is the **sum** of per-dim log-probs (line 278-281), and `entropy` is the **mean** across dims (line 290).

**No new `supervised_loss` method needed on either policy.** The existing NFSP idiom `loss = -log_probs_taken.neg().mean()` (line 827) is precisely sum-of-per-head cross-entropy when the underlying `log_probs_taken` is the multi-discrete summed-across-dims log-prob. Cross-checked against the joint trainer's PPO update at `src/multi_agent/joint.rs:763-786` which uses the identical `evaluate_actions_joint` surface for the PPO surrogate — that's the PR #103 multi-discrete template the issue body references. **No blocker, no architect handoff.**

### 4. Cartesian-product wrapper bijection check

PR #126's `SingleDiscreteBucketBrigade` at `tests/test_nfsp_bucket_brigade.rs:155-189`:

- `FLAT_ACTION_DIM = NUM_HOUSES * 2 * 2 = 10 * 2 * 2 = 40`.
- Decode at lines 181-183: `signal = v % 2; mode = (v / 2) % 2; house = (v / 4) % NUM_HOUSES`.
- This is a faithful bijection over `0..40`: `v = house * 4 + mode * 2 + signal` is the inverse, and the modular ops cancel cleanly when `v < 40`. No information loss.

PR #129's inlined copy at `examples/games/bucket_brigade/train_nfsp.rs:138-172` is byte-identical to the test wrapper (same decode formulas at lines 163-166, same `assert_eq!(a.len(), 1, ...)` assertion). The file header explicitly references #127 at lines 26-31 with a TODO to switch back to the factored policy once this lands.

### 5. Reservoir generalization: recommend Option B (`Vec<i64>`)

The issue body offers two options. **Recommend Option B (`Vec<i64>` action storage)** for these reasons:

1. **No generic-cascade through the trainer.** `NfspTrainer` is already heavily generic (`B, P, O, E, FP, FO, FE`); adding another action-type parameter `A` would propagate through every test builder and example. By contrast, swapping `(Vec<f32>, i64)` → `(Vec<f32>, Vec<i64>)` is a localized type change.
2. **Single-discrete is the degenerate case.** A `Vec<i64>` of length 1 IS a scalar action; storage cost is negligible at reservoir capacity 2000-2_000_000.
3. **Matches the joint trainer's existing convention.** `JointEnv::step_joint` already takes `&[Vec<i64>]` (`src/multi_agent/joint.rs:254`), and `JointRollout::actions` stores actions strided per-dim in a flat `Vec<i64>` with `num_action_dims` metadata. The reservoir naturally aligns with this.
4. **Public API churn is contained.** Only the public type signature at `src/multi_agent/nfsp.rs:489` changes: `&ReservoirBuffer<(Vec<f32>, Vec<i64>)>`.

Option A (generic `A: Clone + Send + Sync`) is more "correct" but pays for flexibility we don't need — every concrete use is `Vec<i64>` actions.

### 6. Supervised-loss reshape

Replace `src/multi_agent/nfsp.rs:799-825` with a flatten that respects `num_action_dims`:

```rust
let num_action_dims = self.br_trainer.policy(i).action_dims_joint().len();
let mut acts_flat = Vec::with_capacity(mb * num_action_dims);
for (_, a) in &batch {
    assert_eq!(a.len(), num_action_dims);
    acts_flat.extend_from_slice(a);
}
let acts_t = Tensor::<B, 2, Int>::from_data(
    TensorData::new(acts_flat, [mb, num_action_dims]),
    &self.device,
);
// No unsqueeze — already shape [mb, num_action_dims].
let (log_probs_taken, _, _) = policy.evaluate_actions_joint(obs_t, acts_t);
let loss = log_probs_taken.neg().mean();
```

The `action_dims_joint()` call on the BR policy is the canonical source of truth (PR #103 / `src/multi_agent/joint.rs:142`). Confirmed it's available because `NfspTrainer.br_trainer` holds the same `P: JointPolicy<B>` modules. (If you'd rather grab from the AP policy instead — both have the same shape after construction; either works.)

### 7. Test/example delta after the fix

Both files exist on `origin/main` and need updating:

- `tests/test_nfsp_bucket_brigade.rs:155-189` — remove `SingleDiscreteBucketBrigade` wrapper; have the test use `BucketBrigadeMaEnv` directly with `MultiDiscreteMlpBurnPolicy` for the BR + AP policies. The `FLAT_ACTION_DIM` constant goes away; the policy is constructed with `action_dims = vec![NUM_HOUSES, 2, 2]`.
- `examples/games/bucket_brigade/train_nfsp.rs:138-172` — same removal. Drop the `#127` comments at lines 14-31, 101-104, 234. Mirror `examples/games/bucket_brigade/train_psro.rs`'s policy construction (PSRO already uses `MultiDiscreteMlpBurnPolicy` directly via PR #103 / #129).

### 8. LOC estimate

Estimated 400-700 LOC of churn:

- `src/multi_agent/nfsp.rs`: ~50-80 LOC (reservoir type, push site, AP train loop).
- New `tests/test_nfsp_multi_discrete_smoke.rs`: ~150-250 LOC (mirrors `test_nfsp_n_player_matching_pennies.rs` shape but with a small multi-discrete env — see the AC for env choice).
- `tests/test_nfsp_bucket_brigade.rs` delta: ~50 LOC removed (the wrapper) + ~50 LOC adjusted (policy construction + eval).
- `examples/games/bucket_brigade/train_nfsp.rs` delta: ~60 LOC removed + ~40 LOC adjusted.
- Updated `nfsp.rs` doc/tests: ~30 LOC.

The Cartesian wrapper itself is ~35 LOC in each file (`SingleDiscreteBucketBrigade` struct + impl + constant), so removing two copies plus the policy-construction adjustment lands cleanly. No upstream policy or `JointPolicy` trait changes required — the existing `MultiDiscreteMlpBurnPolicy` API is sufficient.

### 9. Refined acceptance criteria

- [ ] `ReservoirBuffer` storage parameterized as `(Vec<f32>, Vec<i64>)` in `NfspTrainer` (`src/multi_agent/nfsp.rs:381`). Public accessor `pub fn reservoir(&self, i: usize) -> &ReservoirBuffer<(Vec<f32>, Vec<i64>)>` updated (`src/multi_agent/nfsp.rs:489`).
- [ ] `collect_anticipatory_rollout` reservoir push at `src/multi_agent/nfsp.rs:709` pushes `row.clone()` (the full per-dim action vector), not just `row[0]`.
- [ ] `train_average_policies` (`src/multi_agent/nfsp.rs:774-842`) builds `acts_t: Tensor<B, 2, Int>` of shape `[mb, num_action_dims]` directly — no `unsqueeze_dim` — using `policy.action_dims_joint().len()` as the source of truth for `num_action_dims`.
- [ ] Existing single-discrete regression test `tests/test_nfsp_n_player_matching_pennies.rs` passes unchanged (uses `MlpBurnPolicy`, scalar actions stored as `vec![a]` internally).
- [ ] New `tests/test_nfsp_multi_discrete_smoke.rs` exercises `MultiDiscreteMlpBurnPolicy` end-to-end on a small env (suggestion: a 2-agent variant of `NPlayerMatchingPennies` extended with a dummy second action dim, or `BucketBrigadeMaEnv` with `MAX_ITERATIONS=2` and `ROLLOUT_STEPS=256`). Asserts the trainer runs to completion without panic and that reservoir sizes grow as expected under non-zero η.
- [ ] `tests/test_nfsp_bucket_brigade.rs` `SingleDiscreteBucketBrigade` wrapper (lines 147-189) and `FLAT_ACTION_DIM` constant removed; test constructs `MultiDiscreteMlpBurnPolicy::new(obs_dim, vec![NUM_HOUSES, 2, 2], HIDDEN_DIM, &device)` per agent. The `assert_eq!(acts.len(), 1)` at line 219 becomes `assert_eq!(acts.len(), 3)`.
- [ ] `examples/games/bucket_brigade/train_nfsp.rs` mirror: wrapper removed (lines 128-172), `FLAT_ACTION_DIM` removed (line 105), policy switched to `MultiDiscreteMlpBurnPolicy`. The `#127` references in the module doc (lines 14-31), constant doc (lines 101-104), and tracing line (line 234) updated to reflect the now-native path.
- [ ] `train_nfsp.rs` artifact filename scheme (`nfsp_<cell>_final_{br,ap}_agent<i>.{bin,json}`) preserved; only the underlying policy type changes.
- [ ] No new `supervised_loss` method needed on `MultiDiscreteMlpBurnPolicy` — verified the existing `-mean(log_probs_taken)` idiom in `train_average_policies` is sum-of-per-head cross-entropy under the multi-discrete `evaluate_actions` summed-log-prob contract (`src/policy/multi_discrete_mlp.rs:278-281`).
- [ ] `cargo test --features "training,env-bucket-brigade" --release` passes (the bucket-brigade test is heavy — `--release` is the existing convention).
- [ ] `cargo clippy --all-targets --features "training,env-bucket-brigade" -- -D warnings` clean.
- [ ] `cargo fmt --all -- --check` clean.

### 10. Blocker check

**No architect handoff required.** The fix is mechanical: change one type, push the full vector at one site, build a 2-D tensor instead of a 1-D-then-unsqueeze at another site. The `MultiDiscreteMlpBurnPolicy::evaluate_actions` contract (sum-across-dims log-prob) is exactly what supervised cross-entropy minimization wants for a factored discrete distribution. PR #103 already validated the same surface for PPO; NFSP's AP loss is structurally simpler than PPO's clipped surrogate, so if PPO works on multi-discrete (it does — see `train_psro.rs`), supervised CE will too.

### File references

- `src/multi_agent/nfsp.rs:381,489,709,774-842` — primary surgery sites.
- `src/multi_agent/joint.rs:142,167-218,931-946` — `JointPolicy` impls + `select_actions` template.
- `src/policy/multi_discrete_mlp.rs:253-293` — multi-discrete `evaluate_actions` (consumes `[batch, num_dims]`, returns summed log-prob).
- `src/policy/mlp.rs:360-375` — scalar `evaluate_actions` for reference.
- `tests/test_nfsp_bucket_brigade.rs:147-189,219` — wrapper to remove + assertion to flip.
- `examples/games/bucket_brigade/train_nfsp.rs:14-31,101-105,128-172,234` — wrapper + tracking comments to remove.
- `tests/test_nfsp_n_player_matching_pennies.rs` — regression bar (must keep passing).
- `examples/games/bucket_brigade/train_psro.rs` — the multi-discrete construction template (PR #103/#129 already integrates `MultiDiscreteMlpBurnPolicy` end-to-end).
